src/async_fuse/memfs/s3_metadata.rs

```diff
@@ -543,152 +559,202 @@ impl<S: S3BackEnd + Sync + Send + 'static> MetaData for S3MetaData<S> {
-            Ok((ino, child_type, child_attr)) => (ino, child_type, child_attr),
-            Err(e) => {
-                debug!("lookup() failed to pre-check, the error is: {:?}", e);
-                return Err(e);
-            }
-        };
+        retry_txn!(TXN_RETRY_LIMIT, {
+            let mut txn = self.kv_engine.new_meta_txn().await;
+            let parent_node = self.get_inode_from_txn(txn.as_mut(), parent).await?;
+            let Some(child_entry) = parent_node.get_entry(child_name) else {
+                return build_error_result_from_errno(
+                    Errno::ENOENT,
+                    format!(
+                        "failed to find child name={:?} under parent ino={} and name={:?}",
+                        child_name,
+                        parent,
+                        parent_node.get_name(),
+                    ),
+                );
+            };
 
-        let ttl = Duration::new(MY_TTL_SEC, 0);
-        let child_node = self
-            .get_node_from_kv_engine(child_ino)
-            .await?
-            .ok_or_else(|| build_inconsistent_fs!(child_ino))?;
-        let attr = child_node.lookup_attr();
-        debug!(
-            "ino={} lookup_count={} lookup_attr={:?}",
-            child_ino,
-            child_node.get_lookup_count(),
-            attr
-        );
-        self.set_node_to_kv_engine(child_ino, child_node).await?;
-        let fuse_attr = fs_util::convert_to_fuse_attr(attr);
-        Ok((ttl, fuse_attr, MY_GENERATION))
+            parent_node
+                .get_attr()
+                .check_perm(context.user_id, context.group_id, 1)?;
+
+            let child_ino = child_entry.ino();
+            let child_node = self.get_inode_from_txn(txn.as_mut(), child_ino).await?;
+            let child_attr = child_node.lookup_attr();
+
+            let ttl = Duration::new(MY_TTL_SEC, 0);
+            let fuse_attr = fs_util::convert_to_fuse_attr(child_attr);
+            txn.set(
+                &KeyType::INum2Node(parent),
+                &ValueType::Node(parent_node.into_serial_node()),
+            );
+            txn.set(
+                &KeyType::INum2Node(child_ino),
+                &ValueType::Node(child_node.into_serial_node()),
+            );
+            (txn.commit().await, (ttl, fuse_attr, MY_GENERATION))
+        })
     }
 
     /// Rename helper for exchange rename
     #[instrument(skip(self), err, ret)]
-    #[cfg(feature = "abi-7-23")]
-    async fn rename_exchange_helper(
-        &self,
-        context: ReqContext,
-        param: RenameParam,
-    ) -> DatenLordResult<()> {
+    async fn rename(&self, context: ReqContext, param: RenameParam) -> DatenLordResult<()> {
         let old_parent = param.old_parent;
         let old_name = param.old_name.as_str();
         let new_parent = param.new_parent;
         let new_name = param.new_name.as_str();
         let flags = param.flags;
-        let no_replace = flags == 1; // RENAME_NOREPLACE
```

Comment:
Don't you check the `no_replace` flag? If it's set, the `new_name` (the destination) should not exist.

Reply:
See `exchange`

Reply:
You just checked whether `NOREPLACE` and `EXCHANGE` are both set, but I think you didn't handle flag `NOREPLACE` properly.